pkg/groupmapper/groupmapper.go

```diff
@@ -0,0 +1,207 @@
+// create the groups for a given user based on the provided UserIdentityInfo
+type UserGroupsMapper struct {
+	delegatedUserMapper authapi.UserIdentityMapper
+	groupsClient        userclient.GroupInterface
+	groupsCache         *usercache.GroupCache
+	groupsSynced        func() bool
+}
+
+func NewUserGroupsMapper(delegate authapi.UserIdentityMapper, groupsClient userclient.GroupInterface, groupInformer userinformer.GroupInformer) *UserGroupsMapper {
+	return &UserGroupsMapper{
+		delegatedUserMapper: delegate,
+		groupsClient:        groupsClient,
+		groupsCache:         usercache.NewGroupCache(groupInformer),
+		groupsSynced:        groupInformer.Informer().HasSynced,
+	}
+}
+
+func (m *UserGroupsMapper) UserFor(identityInfo authapi.UserIdentityInfo) (kuser.Info, error) {
+	userInfo, err := m.delegatedUserMapper.UserFor(identityInfo)
+	if err != nil {
+		return userInfo, err
+	}
+
+	identityGroups := sets.NewString(identityInfo.GetProviderGroups()...)
+	if err := m.processGroups(identityInfo.GetProviderName(), identityInfo.GetProviderPreferredUserName(), identityGroups); err != nil {
+		return nil, err
+	}
+
+	return &UserInfoGroupsWrapper{
+		userInfo:         userInfo,
+		additionalGroups: identityGroups,
+	}, nil
+}
+
+func (m *UserGroupsMapper) processGroups(idpName, username string, groups sets.String) error {
+	err := wait.PollImmediate(1*time.Second, 5*time.Second, func() (bool, error) {
+		return m.groupsSynced(), nil
+	})
+	if err != nil {
+		return err
+	}
+
+	cachedGroups, err := m.groupsCache.GroupsFor(username)
+	if err != nil {
+		return err
+	}
+
+	removeGroups, addGroups := groupsDiff(cachedGroups, groups)
+	for _, g := range removeGroups {
+		if err := m.removeUserFromGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	for _, g := range addGroups {
+		if err := m.addUserToGroupWithRetries(idpName, username, g); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// FIXME: work better with annotations
+// FIXME: add retries
+// FIXME: use idpName
+func (m *UserGroupsMapper) removeUserFromGroupWithRetries(idpName, username, group string) error {
+	updatedGroup, err := m.groupsClient.Get(context.TODO(), group, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if len(updatedGroup.Users) == 0 {
+		return nil
+	}
+
+	if len(updatedGroup.Users) == 1 && updatedGroup.Users[0] == username && updatedGroup.Annotations[groupGeneratedKey] == "true" {
+		return m.groupsClient.Delete(context.TODO(), group, metav1.DeleteOptions{})
+	}
+
+	// find the user and remove it from the slice
+	userIdx := -1
+	for i, groupUser := range updatedGroup.Users {
+		if groupUser == username {
+			userIdx = i
+			break
+		}
+	}
+
+	var newUsers []string
+	switch userIdx {
+	case -1:
+		return nil
+	case 0:
+		newUsers = updatedGroup.Users[1:]
+	default:
+		newUsers = append(updatedGroup.Users[0:userIdx], updatedGroup.Users[userIdx+1:]...)
```

Comment:
Can it happen, that the user is the last one? Would it be out of bounds?

Reply:
golang appears to be returning an empty slice on `[tooGreatOfAnIndex:]`

Reply:
I'd rather stick to the smallest number of cases possible and not write code that does not need to exist